contrib/ibmock/efa.c

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2024, NVIDIA CORPORATION. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include <infiniband/efadv.h>
+#include <infiniband/verbs.h>
+
+#include <errno.h>
+#include <stdlib.h>
+
+#include "fake.h"
+#include "verbs.h"
+#include "config.h"
+
+int efadv_query_device(struct ibv_context *context,
+                       struct efadv_device_attr *attr, uint32_t inlen)
+{
+    if (!context || inlen != sizeof(efa_dev_attr)) {
+        return EINVAL;
+    }
+
+    if (context->device->node_type != IBV_NODE_UNSPECIFIED ||
+        context->device->transport_type != IBV_TRANSPORT_UNSPECIFIED) {
+        return ENOTSUP;
+    }
+
+    memcpy(attr, &efa_dev_attr, sizeof(efa_dev_attr));
+    return 0;
+}
+
+struct ibv_qp *efadv_create_driver_qp_impl(struct ibv_pd           *pd,
+                                           struct ibv_qp_init_attr *attr,
+                                           uint32_t driver_qp_type)
+{
+    struct fake_qp *fqp = calloc(1, sizeof(*fqp));
+    struct ibv_qp *qp   = &fqp->qp_ex.qp_base;
```

Comment:
maybe do this assignment after check for NULL?

Reply:
fixed, others too thanks